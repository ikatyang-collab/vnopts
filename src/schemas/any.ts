import { Schema, SchemaParameters } from '../schema.js'
import { ValidateResult } from '../types.js'

export class AnySchema extends Schema<any, SchemaParameters<any>> {
  public expected() {
    return 'anything'
  }
  public validate(): ValidateResult {
    return true
  }
}
