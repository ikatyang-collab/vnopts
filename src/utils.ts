import {
  DefaultResult,
  DeprecatedResult,
  ForwardResult,
  InvalidHandler,
  NormalizedDefaultResult,
  NormalizedDeprecatedResult,
  NormalizedDeprecatedResultWithTrue,
  NormalizedForwardResult,
  NormalizedInvalidHandler,
  NormalizedRedirectResult,
  NormalizedTransferResult,
  NormalizedValidateResult,
  Options,
  RedirectResult,
  TransferResult,
  ValidateResult,
} from './types';

export type NotEmptyArray<T> = [T, ...T[]];

export type RecordFromUnion<T extends Options, K extends keyof T> = {
  [X in T[K]]: Extract<T, Record<K, X>>
};

export function recordFromArray<
  T extends Record<string, any>,
  K extends keyof T
>(array: T[], mainKey: K): RecordFromUnion<T, K> {
  const record = Object.create(null) as Partial<RecordFromUnion<T, K>>;

  for (const value of array) {
    const key = value[mainKey];

    // istanbul ignore next
    if (record[key]) {
      throw new Error(`Duplicate ${mainKey} ${JSON.stringify(key)}`);
    }

    // @ts-ignore
    record[key] = value;
  }

  return record as RecordFromUnion<T, K>;
}

export function mapFromArray<T extends Record<string, any>, K extends keyof T>(
  array: T[],
  mainKey: K,
): Map<T[K], T> {
  const map = new Map<T[K], T>();

  for (const value of array) {
    const key = value[mainKey];

    // istanbul ignore next
    if (map.has(key)) {
      throw new Error(`Duplicate ${mainKey} ${JSON.stringify(key)}`);
    }

    map.set(key, value);
  }

  return map;
}

export function createAutoChecklist() {
  const map: { [idString: string]: boolean } = Object.create(null);
  return (id: any) => {
    const idString = JSON.stringify(id);
    if (map[idString]) {
      return true;
    }
    map[idString] = true;
    return false;
  };
}

export function partition<T>(
  array: T[],
  predicate: (x: T) => boolean,
): [T[], T[]] {
  const trueArray: T[] = [];
  const falseArray: T[] = [];

  for (const value of array) {
    if (predicate(value)) {
      trueArray.push(value);
    } else {
      falseArray.push(value);
    }
  }

  return [trueArray, falseArray];
}

export function isInt(value: number) {
  return value === Math.floor(value);
}

export function comparePrimitive(
  a: undefined | null | boolean | number | string,
  b: undefined | null | boolean | number | string,
): number {
  if (a === b) {
    return 0;
  }

  const typeofA = typeof a;
  const typeofB = typeof b;

  const orders = [
    'undefined',
    'object', // null
    'boolean',
    'number',
    'string',
  ];
  if (typeofA !== typeofB) {
    return orders.indexOf(typeofA) - orders.indexOf(typeofB);
  }

  if (typeofA !== 'string') {
    return Number(a) - Number(b);
  }

  return (a as string).localeCompare(b as string);
}

export function normalizeInvalidHandler(
  invalidHandler: InvalidHandler,
): NormalizedInvalidHandler {
  return (...args) => {
    const errorMessageOrError = invalidHandler(...args);
    return typeof errorMessageOrError === 'string'
      ? new Error(errorMessageOrError)
      : /* istanbul ignore next*/ errorMessageOrError;
  };
}

export function normalizeDefaultResult<$Value>(
  result: DefaultResult<$Value>,
): NormalizedDefaultResult<$Value> {
  return result === undefined ? {} : result;
}

export function normalizeValidateResult(
  result: ValidateResult,
  value: unknown,
): NormalizedValidateResult {
  return result === true ? true : result === false ? { value } : result;
}

export function normalizeDeprecatedResult<$Value>(
  result: DeprecatedResult<$Value>,
  value: $Value,
): NormalizedDeprecatedResult<$Value>;
export function normalizeDeprecatedResult<$Value>(
  result: DeprecatedResult<$Value>,
  value: $Value,
  doNotNormalizeTrue: true,
): NormalizedDeprecatedResultWithTrue<$Value>;
export function normalizeDeprecatedResult<$Value>(
  result: DeprecatedResult<$Value>,
  value: $Value,
  doNotNormalizeTrue = false,
):
  | NormalizedDeprecatedResult<$Value>
  | NormalizedDeprecatedResultWithTrue<$Value> {
  return result === false
    ? false
    : result === true
      ? doNotNormalizeTrue
        ? true
        : [{ value }]
      : 'value' in result
        ? [result]
        : result.length === 0
          ? false
          : (result as NotEmptyArray<{ value: $Value }>);
}

export function normalizeTransferResult<$Value>(
  result: TransferResult<$Value>,
  value: $Value,
): NormalizedTransferResult<$Value> {
  return typeof result === 'string' || 'key' in result
    ? { from: value, to: result }
    : 'from' in result
      ? { from: result.from!, to: result.to }
      : { from: value, to: result.to };
}

export function normalizeForwardResult<$Value>(
  result: ForwardResult<$Value>,
  value: $Value,
): NormalizedForwardResult<$Value> {
  return result === undefined
    ? []
    : Array.isArray(result)
      ? result.map(transferResult =>
          normalizeTransferResult(transferResult, value),
        )
      : [normalizeTransferResult(result, value)];
}

export function normalizeRedirectResult<$Value>(
  result: RedirectResult<$Value>,
  value: $Value,
): NormalizedRedirectResult<$Value> {
  const redirect = normalizeForwardResult(
    typeof result === 'object' && 'redirect' in result
      ? result.redirect
      : result,
    value,
  );
  return redirect.length === 0
    ? { remain: value, redirect }
    : typeof result === 'object' && 'remain' in result
      ? { remain: result.remain, redirect }
      : { redirect };
}
