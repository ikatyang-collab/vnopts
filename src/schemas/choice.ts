import { Schema, SchemaHandlers, SchemaParameters } from '../schema';
import {
  DeprecatedResult,
  ExpectedResult,
  ForwardResult,
  RedirectResult,
  TransferTo,
  Utils,
  ValidateResult,
} from '../types';
import { comparePrimitive, mapFromArray } from '../utils';

interface ChoiceInfo {
  value: ChoiceValue;
  deprecated?: boolean;
  redirect?: TransferTo;
  forward?: TransferTo;
  /** do not show this value in `expected` */
  hidden?: boolean;
}

type ChoiceValue = undefined | null | boolean | number | string;

interface ChoiceSchemaParameters
  extends SchemaParameters<ChoiceValue>,
    SchemaHandlers<ChoiceValue> {
  choices: Array<ChoiceValue | ChoiceInfo>;
}

export class ChoiceSchema extends Schema<ChoiceValue, ChoiceSchemaParameters> {
  private _choices: Map<ChoiceValue, ChoiceInfo>;

  constructor(parameters: ChoiceSchemaParameters) {
    super(parameters);
    this._choices = mapFromArray(
      parameters.choices.map(
        choice =>
          choice && typeof choice === 'object' ? choice : { value: choice },
      ),
      'value',
    );
  }

  public expected({ descriptor }: Utils): ExpectedResult {
    const choiceDescriptions = Array.from(this._choices.keys())
      .map(value => this._choices.get(value)!)
      .filter(({ hidden }) => !hidden)
      .map(choiceInfo => choiceInfo.value)
      .sort(comparePrimitive)
      .map(descriptor.value);

    const head = choiceDescriptions.slice(0, -2);
    const tail = choiceDescriptions.slice(-2);
    const message = head.concat(tail.join(' or ')).join(', ');

    return {
      text: message,
      list: {
        title: 'one of the following values',
        values: choiceDescriptions,
      },
    };
  }

  public validate(value: unknown): ValidateResult {
    return this._choices.has(value as any);
  }

  public deprecated(value: ChoiceValue): DeprecatedResult<ChoiceValue> {
    const choiceInfo = this._choices.get(value);
    return choiceInfo && choiceInfo.deprecated ? { value } : false;
  }

  public forward(value: ChoiceValue): ForwardResult<ChoiceValue> {
    const choiceInfo = this._choices.get(value);
    return choiceInfo ? choiceInfo.forward : undefined;
  }

  public redirect(value: ChoiceValue): RedirectResult<ChoiceValue> {
    const choiceInfo = this._choices.get(value);
    return choiceInfo ? choiceInfo.redirect : undefined;
  }
}
