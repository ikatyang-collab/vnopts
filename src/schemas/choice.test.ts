import { eachHandler } from '../../tests/__helpers__/utils';
import { AnySchema } from './any';
import { ChoiceSchema } from './choice';

const name = '<key>';
const value = '<value>';

const transferKey = '<transfer-key>';
const transferValue = '<transfer-value>';
const transferPair = { key: transferKey, value: transferValue };
const transfer = { [transferKey]: transferValue };

const invalidValue = '<invalid-value>';
const deprecatedValue = '<deprecated-value>';
const forwardValue = '<forward-value>';
const redirectValue = '<redirect-value>';
const unknownValue = '<unknown-value>';
const hiddenValue = '<hidden-value>';

describe.each`
  parameters            | input                          | output                                   | hasWarnings
  ${{}}                 | ${{ [name]: value }}           | ${{ [name]: value }}                     | ${false}
  ${{}}                 | ${{ [name]: invalidValue }}    | ${Error}                                 | ${false}
  ${{}}                 | ${{ [name]: deprecatedValue }} | ${{ [name]: deprecatedValue }}           | ${true}
  ${{}}                 | ${{ [name]: forwardValue }}    | ${{ [name]: forwardValue, ...transfer }} | ${false}
  ${{}}                 | ${{ [name]: redirectValue }}   | ${transfer}                              | ${false}
  ${{ validate: true }} | ${{ [name]: unknownValue }}    | ${{ [name]: unknownValue }}              | ${false}
`(
  '',
  eachHandler<ChoiceSchema>(
    ChoiceSchema,
    {
      name,
      choices: [
        value,
        { value: deprecatedValue, deprecated: true },
        { value: forwardValue, forward: transferPair },
        { value: redirectValue, redirect: transferPair },
        { value: hiddenValue, hidden: true },
      ],
    },
    { schemas: [new AnySchema({ name: transferKey })] },
  ),
);

for (const loggerPrintWidth of [-Infinity, Infinity]) {
  describe.each`
    parameters | input                    | output   | hasWarnings
    ${{}}      | ${{ [name]: 'invalid' }} | ${Error} | ${false}
  `(
    `(expected values printWidth=${loggerPrintWidth})`,
    eachHandler<ChoiceSchema>(
      ChoiceSchema,
      {
        name,
        choices: Array.from(new Array(5), (_value, index) => `value${index}`),
      },
      { loggerPrintWidth },
    ),
  );
}
