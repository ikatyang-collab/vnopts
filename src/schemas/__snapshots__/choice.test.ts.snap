// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`(expected values printWidth=-Infinity) schema {"name":"<key>","choices":["value0","value1","value2","value3","value4"]} with input {"<key>":"invalid"} 1`] = `
"Invalid \\"<key>\\" value. Expected one of the following values, but received \\"invalid\\".
- \\"value0\\"
- \\"value1\\"
- \\"value2\\"
- \\"value3\\"
- \\"value4\\""
`;

exports[`(expected values printWidth=Infinity) schema {"name":"<key>","choices":["value0","value1","value2","value3","value4"]} with input {"<key>":"invalid"} 1`] = `"Invalid \\"<key>\\" value. Expected \\"value0\\", \\"value1\\", \\"value2\\", \\"value3\\" or \\"value4\\", but received \\"invalid\\"."`;

exports[`schema {"name":"<key>","choices":["<value>",{"value":"<deprecated-value>","deprecated":true},{"value":"<forward-value>","forward":{"key":"<transfer-key>","value":"<transfer-value>"}},{"value":"<redirect-value>","redirect":{"key":"<transfer-key>","value":"<transfer-value>"}}]} with input {"<key>":"<deprecated-value>"} 1`] = `
Array [
  "{ \\"<key>\\": \\"<deprecated-value>\\" } is deprecated.",
]
`;

exports[`schema {"name":"<key>","choices":["<value>",{"value":"<deprecated-value>","deprecated":true},{"value":"<forward-value>","forward":{"key":"<transfer-key>","value":"<transfer-value>"}},{"value":"<redirect-value>","redirect":{"key":"<transfer-key>","value":"<transfer-value>"}}]} with input {"<key>":"<invalid-value>"} 1`] = `
"Invalid \\"<key>\\" value. Expected one of the following values, but received \\"<invalid-value>\\".
- \\"<forward-value>\\"
- \\"<redirect-value>\\"
- \\"<value>\\""
`;
